### PR TITLE
added support for nil values

### PR DESCRIPTION
**Pull request by woutslakhorst**
Don't know if it's wise to merge this.....

Always think out-of-the-box and a better solution presents itself!

### CODE REVIEW

**Comment by woutslakhorst** [comment]
To really support this, it might be better to change the serialized `Entry`'s in the bbolt value to a bucket reference when things grow too big.....

**Comment by woutslakhorst** [comment]
A quick performance test shows that subbuckets are faster when more than 16 items are placed under the same key. The performance benefit grows exponentially with the number of items per key. The size of the db is larger for the subbucket scenario 2 times the size (with a larger fill percentage).

**Comment by woutslakhorst** [comment]
not backwards compatible though

File: index.go [hunk 1]
```diff
@@ -126,6 +127,10 @@ func addDocumentR(bucket *bbolt.Bucket, parts []FieldIndexer, cKey Key, ref Refe
 			key := ComposeKey(cKey, m)
 			_ = addRefToBucket(bucket, key, ref)
 		}
+		if len(matches) == 0 {
```
Comment (reinkrul) [inline index.go:130]: I dont understand this "if"?

File: index.go [hunk 2]
```diff
@@ -181,50 +192,24 @@ func (i *index) Delete(bucket *bbolt.Bucket, ref Reference, doc Document) error
...
-	}
-	entry.Delete(ref)
-
-	if entry.Size() == 0 {
-		return bucket.Delete(key)
-	}
-
-	iBytes, err := entry.Marshal()
+	// first check if there's a sub-bucket
+	subBucket, err := bucket.CreateBucketIfNotExists(key)
```
Comment (reinkrul) [inline index.go:207]: why create the bucket if the item to remove doesnt exist?

File: index.go [hunk 3]
```diff
@@ -181,50 +192,24 @@ func (i *index) Delete(bucket *bbolt.Bucket, ref Reference, doc Document) error
...
 	if err != nil {
 		return err
 	}
-
-	return bucket.Put(key, iBytes)
+	subBucket.FillPercent = 0.9
```
Comment (reinkrul) [inline index.go:200]: why 0.9?

File: index.go [hunk 4]
```diff
@@ -181,50 +192,24 @@ func (i *index) Delete(bucket *bbolt.Bucket, ref Reference, doc Document) error
...
+	return subBucket.Put(ref, []byte{})
```
Comment (reinkrul) [inline index.go:201]: I don't see the part which limits the values per entry (and creates another subbucket), from what I understood from your explanation?

File: index.go [hunk 5]
```diff
@@ -372,8 +357,14 @@ func findR(cursor *bbolt.Cursor, sKey Key, matchers []matcher, fn iteratorFn) er
 					nKey := ComposeKey(sKey, newp)
 					err = findR(cursor, nKey, matchers[1:], fn)
 				} else {
-					// value found call iterator function
-					err = fn(cKey, entry)
+					// sub-bucket found, call iterator function
+					subBucket := cursor.Bucket().Bucket(cKey)
+					if subBucket != nil {
+						subCursor := subBucket.Cursor()
+						for k, _ := subCursor.Seek([]byte{}); k != nil; k, _ = subCursor.Next() {
```
Comment (reinkrul) [inline index.go:364]: the cognitive and cyclomatic complexity here has just entered orbit... :'-)

**Review by reinkrul — COMMENTED** [review]
Need some more explanation to have a well-founded opinion on this.

File: index.go (hunk 4 above)
Comment (woutslakhorst) [reply index.go:201]: Everything is in the subbucket now. No more entries in the value. Subbuckets can grow until the max DB size.

File: index.go (hunk 3 above)
Comment (woutslakhorst) [reply index.go:200]: I removed those percentages, did some testing and it doesn't influence size of the DB significantly.

File: index.go (hunk 2 above)
Comment (woutslakhorst) [reply index.go:207]: changed it to a bucket check, if no bucket return nil.

File: index.go (hunk 1 above)
Comment (woutslakhorst) [reply index.go:130]: This means that the JSON path query for this index part didn't return any result => a nil value. A nil value is reflected in the compound key with two adjacent separators:

for an index with 3 indexed fields, where the third part resulted in a nil value
```key = value1 + 0x10 + value2 + 0x10 + 0x10```

File: index.go (hunk 5 above)
Comment (woutslakhorst) [reply index.go:364]: simplified the document retrieval, also noticed that the `cursor.seek` in the `else` wasn't even needed.

**Review by reinkrul — APPROVED** [review]

**Pull request merged**